### PR TITLE
added functions and model for out-of-sample prediction

### PR DESCRIPTION
**Pull request by IanEisenberg**
WIP addresses #12 

Current issues that need to be addressed:

- Prediction model compiles every time `fit_TIRTpredictor_stan` is called. Should be precompiled somewhere.
- Only works with "gamma", not "gamma_ord" so doesn't apply to all functional families.

### CODE REVIEW

**Comment by paul-buerkner** [comment]
Thank you very much! I don't have time this week to look at this in detail but will try to come back to it next week.

**Comment by IanEisenberg** [comment]
Will this be edited/incorporated into the package at any point in the near future? Maybe early in the new decade?

**Comment by paul-buerkner** [comment]
I will try to work on that in the first week of january. Sorry for the
delay but I need to edit this a little bit and had to prioritize other
things first.

Ian <notifications@github.com> schrieb am Fr., 20. Dez. 2019, 23:50:

File: R/stan.R [hunk 1]
```diff
@@ -153,3 +153,50 @@ fit_TIRT_stan <- function(data, init = 0, ...) {
...
+  out = list(lambda = fit_lambda,
+             phi = fit_Phi,
+             gamma = fit_gamma,
+             psi = fit_psi,
+             eta = fit_eta,
+             r = fit_r)
+  return(out)
+}
+
+get_prediction_model <- function(fitted_model) {
```
Comment (paul-buerkner) [inline R/stan.R]: I am a little confused about the purpose of the function. Specifically, there seems to be a local path involved. Is this function really necessary or can we just copy the code over to `fit_TIRTpredictor_stan`?

File: R/stan.R [hunk 2]
```diff
@@ -153,3 +153,50 @@ fit_TIRT_stan <- function(data, init = 0, ...) {
   structure(nlist(fit, data), class = "TIRTfit")
 }
 
+#' @export
```
Comment (paul-buerkner) [inline R/stan.R]: extracts mean components from all parameters but how would we handle the case when we want to whole posterior distribution of the new trait scores not just their means?

File: R/stan.R [hunk 3]
```diff
@@ -153,3 +153,50 @@ fit_TIRT_stan <- function(data, init = 0, ...) {
...
+get_TIRT_params_stan <- function(fitted_model) {
+  # extract fit parameters from the stanmodel object
+  components = rstan::extract(fitted_model)
+  # person parameters
+  fit_eta = apply(components$eta,c(2,3),mean)
+  colnames(fit_eta) = sprintf("trait%s",seq(1:ncol(fit_eta)))
```
Comment (paul-buerkner) [inline R/stan.R]: Do we need the old eta values in the prediction of new data?

File: src/stan_files/thurstonian_irt_predictor.stan [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  int<lower=1> J_item_orig[N_item_equal];
+  // number of response categories in ordinal models
+  // should be set to 2 for other families
+  int<lower=2> ncat;
+}
+transformed data {
+  vector<lower=0>[N_item_fix] psi_fix;  // fixed item SDs
+  psi_fix = rep_vector(1.0, N_item_fix);
+}
+parameters {
```
Comment (paul-buerkner) [inline src/stan_files/thurstonian_irt_predictor.stan]: What did you change from the original model fitting stan file and the predictor file? I am a little confused as to how many parameters are included in the model here. Wouldn't the only new parameters be the new trait scores?

File: NAMESPACE [hunk 5]
```diff
@@ -11,6 +11,8 @@ export(empty_block)
 export(fit_TIRT_lavaan)
 export(fit_TIRT_mplus)
 export(fit_TIRT_stan)
+export(get_TIRT_params_stan)
```
Comment (paul-buerkner) [inline NAMESPACE]: How would you recommend we integrate this functionality into the `predict` method?

**Comment by paul-buerkner** [comment]
I have addded out-of-sample predictions independently of this PR and will thuse close the PR.

**Pull request closed without merging**